### ISSUE TITLE
no symmetry in graph_edit_distance

### ISSUE THREAD

**Issue by sheldonkhall** [networkx/networkx#5400]
I have been trying to use the network x implementation of the graph edit distance to do some clustering. However this led to some unexpected errors. The root cause of these errors has been a non-symmetric distance matrix that I have tracked back to non-symmetric edit distances produced by the `nx.graph_edit_distance` measure.

A minimal example is this:

g1 is the graph  c - d  with a self edge on d
g2 is the graph e - d - c with no self edges

```
import networkx as nx

n1 = [("c", {"attr": "c"}), ("d", {"attr": "d"})]
n2 = [("e", {"attr": "e"}), ("d", {"attr": "d"}), ("c", {"attr": "c"})]
e1 = [("c", "d", {"attr": {"c", "d"}}), ("d", "d", {"attr": {"d"}})]
e2 = [("c", "d", {"attr": {"c", "d"}}), ("d", "e", {"attr":{"d", "e"}})]

g1 = nx.Graph()
g1.add_nodes_from(n1)
g1.add_edges_from(e1)
g2 = nx.Graph()
g2.add_nodes_from(n2)
g2.add_edges_from(e2)

match = lambda x, y: x["attr"] == y["attr"]
print(nx.graph_edit_distance(g1, g2, node_match=match, edge_match=match) == nx.graph_edit_distance(g2, g1, node_match=match, edge_match=match))
```

This returns False where I would expect it to be True. I am using the node and edge match methods because I want there to be a substitution cost for nodes and edges.

Can anyone confirm this behaviour is a bug?

In the meantime I will investigate further and check my assumptions.

**Comment by dschult** [comment]
My understanding is that the graph_edit_distance can depend on the order of the inputs if the match function causes it to.
For example, g1->g2 has cost 3 with steps: add node "e", remove edge ("d","d"), add edge ("d","e").
While g2->g1 has cost 2 with steps: rewire edge ("d","e") to ("d","d"), remove node "e".

In other words with this match function, rewiring ("d","d") to ("d","e") takes two steps, while ("d","e") to ("d","d") takes one step.  I believe this is because of the self-loop attribute representation of a set of one value -- rather than a tuple of two identical values, but I'm not actually sure.

Removing the match function certainly makes it work symmetrically. using a match that is always True also removes this difference.  Does this help any?

**Comment by sheldonkhall** [comment]
The match function I have used is designed to only match when the undirected edge is the same. Therefore as you say ("d","d") will not match ("d", "e"). The surprising behaviour is that instead of picking a substitution ("d", "d") -> ("d", "e") for a cost of 1 it opts for a delete and then an insert which has a cost of 2. This is not the minimum edit path as it has a total cost of 3. However, in the opposite direction it does pick the minimum edit path of 2.

The story gets more complicated though. The example below is exactly the same but I have changed the order I add the nodes `n2`. In this example I add the nodes to `g2` in the order ( c - d - e ) which is an identical graph. However, in this case the edit paths are both 2, which is what I would expect.

```
n1 = [("c", {"attr": "c"}), ("d", {"attr": "d"})]
n2 = [("c", {"attr": "c"}), ("d", {"attr": "d"}), ("e", {"attr": "e"})]
e1 = [("c", "d", {"attr": {"c", "d"}}), ("d", "d", {"attr": {"d"}})]
e2 = [("c", "d", {"attr": {"c", "d"}}), ("d", "e", {"attr":{"d", "e"}})]

g1 = nx.Graph()
g1.add_nodes_from(n1)
g1.add_edges_from(e1)
g2 = nx.Graph()
g2.add_nodes_from(n2)
g2.add_edges_from(e2)

match = lambda x, y: x["attr"] == y["attr"]
print(nx.graph_edit_distance(g1, g2, node_match=match, edge_match=match) == nx.graph_edit_distance(g2, g1, node_match=match, edge_match=match))
```

Is it correct that adding nodes in a different order to produce an identical graph should result in a different minimum edit path? That is what is making me wonder if there is a bug.

**Comment by dschult** [comment]
Seems like node order does matter for this code.  It was written before dict's maintained insertion order, so this could be a subtle bug that got created during that shift. It only seems to matter for self-loops making it even harder to identify.  Thank you for tracking it down to such simple input graphs!  Now we just need to find where in the code we treat self-loops.

Perhaps @aparamon will see this and come up with a brilliant quick fix. :}

Following the discussion in #2361 I uncommented all the print_debug statements and ran the resulting code with:
```python
print("G2 is edge (a,b) and G3 is edge (a,a)")
print("but node order for G2 is (a,b) while for G3 it is (b,a)")

a, b = "A", "B" 
G2 = nx.Graph()
G2.add_nodes_from((a, b)) 
G2.add_edges_from([(a, b)])
G3 = nx.Graph()
G3.add_nodes_from((b, a)) 
G3.add_edges_from([(a, a)])
for G in (G2, G3):
    for n in G:
        G.nodes[n]["attr"] = n 
    for e in G.edges:
        G.edges[e]["attr"] = e 
match = lambda x, y: x == y

print("Starting G2 to G3 GED calculation")
assert graph_edit_distance(G2, G3, node_match=match, edge_match=match) == 2

print("Starting G3 to G2 GED calculation")
assert graph_edit_distance(G3, G2, node_match=match, edge_match=match) == 2
```
This is the output
```
G2 is edge (a,b) and G3 is edge (a,a)
but node order for G2 is (a,b) while for G3 it is (b,a)
Starting G2 to G3 GED calculation
Cv: 2 x 2
[[1. 0. 1. 7.]
 [0. 1. 7. 1.]
 [1. 7. 0. 0.]
 [7. 1. 0. 0.]]
Ce: 1 x 1
[[1. 1.]
 [1. 0.]]

matched-uv: []
matched-gh: []
matched-cost: 0
pending-u: ['A', 'B']
pending-v: ['B', 'A']
[[1. 0. 1. 7.]
 [0. 1. 7. 1.]
 [1. 7. 0. 0.]
 [7. 1. 0. 0.]]
pending-g: [('A', 'B')]
pending-h: [('A', 'A')]
[[1. 1.]
 [1. 0.]]

matched-uv: [('A', 'A')]
matched-gh: [(None, ('A', 'A'))]
matched-cost: 1.0
pending-u: ['B']
pending-v: ['B']
[[0. 1.]
 [1. 0.]]
pending-g: [('A', 'B')]
pending-h: []
[[1.]]

matched-uv: [('A', 'A'), ('B', 'B')]
matched-gh: [(None, ('A', 'A')), (('A', 'B'), None)]
matched-cost: 2.0
pending-u: []
pending-v: []
[]
pending-g: []
pending-h: []
[]

Starting G3 to G2 GED calculation
Cv: 2 x 2
[[1. 0. 1. 7.]
 [0. 1. 7. 1.]
 [1. 7. 0. 0.]
 [7. 1. 0. 0.]]
Ce: 1 x 1
[[1. 1.]
 [1. 0.]]

matched-uv: []
matched-gh: []
matched-cost: 0
pending-u: ['B', 'A']
pending-v: ['A', 'B']
[[1. 0. 1. 7.]
 [0. 1. 7. 1.]
 [1. 7. 0. 0.]
 [7. 1. 0. 0.]]
pending-g: [('A', 'A')]
pending-h: [('A', 'B')]
[[1. 1.]
 [1. 0.]]

matched-uv: [('B', 'B')]
matched-gh: []
matched-cost: 0.0
pending-u: ['A']
pending-v: ['A']
[[0. 1.]
 [1. 0.]]
pending-g: [('A', 'A')]
pending-h: [('A', 'B')]
[[1. 1.]
 [1. 0.]]

matched-uv: [('B', 'B'), ('A', 'A')]
matched-gh: [(('A', 'A'), ('A', 'B'))]
matched-cost: 1.0
pending-u: []
pending-v: []
[]
pending-g: []
pending-h: []
[]

AssertionError:  1.0 <> 2
```

**Comment by sheldonkhall** [comment]
Thanks for simplifying the test further @dschult . I will spend a little time this week understanding the code in more detail and see if I can spot a fix while we wait to see if @aparamon has any ideas.

**Comment by sheldonkhall** [comment]
@dschult I have isolated the lines causing this behaviour: 
- https://github.com/networkx/networkx/blob/1e48dad5c6038b549d7a4a01f190ffd3765eb913/networkx/algorithms/similarity.py#L751 
- https://github.com/networkx/networkx/blob/1e48dad5c6038b549d7a4a01f190ffd3765eb913/networkx/algorithms/similarity.py#L757

On those lines there is a specific condition for self edges. When that gets triggered for some reason an edge substitution is not picked as the correct path and it picks the edge delete.

**Comment by dschult** [comment]
I agree that those two lines look for selfloops.  But it's not clear to me whether the next line might also get triggered sometimes when there is a selfloop (that is not simply u-u or v-v).  Then there is tracking down what `ind_h` and `_ind_g` get used for.  Presumably they flag those edges to be changed -- but it is in the logic which tries to change them that we will find whether an edge substitution or an edge delete is chosen.

Thanks for this --

**Comment by sheldonkhall** [comment]
@dschult I have a working fix now [here](sheldonkhall:bugfix-for-issue-5400). I think it is not ready to merge quite yet as I want to check the one failing assert statement and I am not sure if my fix is entirely general just yet.

**Comment by rossbar** [comment]
I believe this was closed by #5516 - please feel free to reopen if I'm mistaken!
